guest-c: Reformatting, formatting flags

guybedford opened the pull request:
This updates the C generator to instead of emitting only two source sections - `h` and `c`, instead to emit multiple subsections of these files - `h_defs`, `h_fns`, `h_helpers`, `adapter_fns`, `c_includes`, `c_fns`, `c_helpers`.

The result is that final emission groups these sections, keeping definitions, functions, exports and imports, helpers, adapters and includes together. Comments are added to the tops of each section as needed.

In addition, boilerplate functions are created for exported functions, with `// ...` comments for filling in, to create a starting point for implementation.

Some output flags are added to the options:
* `--no-boilerpate`: Skips emitting the new `.c` boilerplate functions. This is also necessary to be the default when not in CLI for the internal tests to pass.
* `--no-helpers`: Skips emitting the user helper functions that aren't called by the generated code.
* `--self-contained-header, -s`: Emits the adapter functions into the header file instead of the c file. This allows a complete separation of user code into the `.c` file and generated bindings into the `.h` file.
* `--no-c`: Does not emit a `.c` file at all. When used with `-s`, this implies a workflow whereby `.wit -> .h` can be an automated update process, with `.c` files being managed manually (including any adjustments to post exec or allocation handling).

There was also a bug in the imports and exports output, whereby exports output was being repeated in the imports output. This is fixed with a new reset line - `self.src = Default::default()` after the emission. There are still some issues at this level that as far as I can tell already existed - ideally I think emission should not separate imports and exports now that we have world files, and emit a single world for both imports and exports, but do let me know if I've understood that part correctly or not.

Minor whitespace and formatting adjustments are also included. Mostly just a bunch of ideas - feedback welcome.

alexcrichton reviewed (commented):
Thanks for this! After reading this over though I'm not sure that this may be the best way to structure this. Reading over these options I feel like they can be combined/used in somewhat confusing ways and additionally lead to possible idioms/usage that might be confusing as well. I think the purpose of these changes though are great and I think some reorganization could help.

Some things I've noticed are:

* Passing just `--no-c` feels not too-useful because then the generated header can't really be used since nothing will define the needed symbols.
* The `__wasm_import_*` functions were previously generated in the "private implementation" `*.c` file since they're not supposed to be called directly, but now they're in the public header which may mislead users into using them instead of the shims which translate.
* I'm not sure of the intended use of the `-s` option since it generates a header that can only be `#include`'d once, which feels not too useful compared to the current generation of things today.
* Generating the boilerplate code with `// ...` comments I feel is useful, but having them intermingled in an otherwise auto-generated `*.c` file feels wrong. I would expect the boilerplate functions to be in a separate `*.c` file that's marked "you should edit this" where other files are ideally marked "please don't edit this" (e.g. https://github.com/bytecodealliance/wit-bindgen/issues/288)

One possible option here is to largely redo how C generation works. Right now for imported functions an "idiomatic C" shim is defined which translates from "idiomatic C" to the canonical ABI. Exports have a shim generated which takes the canonical ABI and calls an "idiomatic C" function. That being said "idiomatic C" AFAIK isn't really a thing per-se, so we could do away with all these shims entirely and instead simply just generate a header file with the canonical ABI and type definitions. There's not necessarily an inherent need to have a lot of "niceties" when working with C as is typically expected with other languages.

With such a scheme I think some of your options and goals may make more sense, such as:

* The header files is "just" types and import/export declarations, no extra fluff. This can be organized as-needed but it's suitable to just drop header files in a project, start calling them, and you're good to go.
* The `*.c` file generate is "just" optional boilerplate for exports. Otherwise for imports nothing is necessary except perhaps a `realloc.c` to generate `cabi_realloc` or something like that.

This is a more radical change and would be more invasive, but `wit-bindgen` isn't in a position where it needs to worry about compatibility to a large degree or anything like that, so I think it would be reasonable to have such changes.

guybedford commented on crates/gen-guest-c/src/lib.rs:
```diff
@@ -44,10 +46,28 @@ struct Func {
+    #[cfg_attr(feature = "clap", arg(long))]
+    no_c: bool,
+    #[cfg_attr(feature = "clap", arg(long))]
+    no_boilerplate: bool,
```
Okay, there's only one flag now:

```
    /// Skip emitting component allocation helper functions
    #[cfg_attr(feature = "clap", arg(long))]
    no_helpers: bool,
```

guybedford commented:
Thanks for the feedback, I've removed the unnecessary flags and boilerplate approaches for C, which agreed do seem unnecessary. Instead, there's now only a `--no-helpers` flag, and I've added weak attributes on the post return functions so that overrides are possible without editing the code in line with the primary intentions.

alexcrichton reviewed (approved):
Sounds good to me!
